crates/aptos-in-memory-cache/src/caches/fifo.rs

```diff
@@ -0,0 +1,426 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Cache, OrderedCache, StreamableOrderedCache};
+use dashmap::DashMap;
+use futures::{stream, Stream};
+use parking_lot::RwLock;
+use std::{fmt::Debug, hash::Hash, sync::Arc};
+use tokio::sync::Notify;
+use tokio_util::sync::CancellationToken;
+use tracing::info;
+
+#[derive(Debug, Clone, Copy)]
+struct CacheMetadata<K> {
+    eviction_trigger_size_in_bytes: u64,
+    target_size_in_bytes: u64,
+    total_size_in_bytes: u64,
+    last_key: Option<K>,
+    first_key: Option<K>,
+}
+
+/// A simple in-memory cache with a deterministic FIFO eviction policy.
+pub struct FIFOCache<K, V>
+where
+    K: Hash + Eq + PartialEq + Send + Sync + Clone + 'static,
+    V: Send + Sync + Clone + 'static,
+{
+    /// Precache to support out of order inserts
+    precache: Arc<DashMap<K, V>>,
+    precache_insert_notify: Arc<Notify>,
+    /// Cache maps the cache key to the deserialized Transaction.
+    items: Arc<DashMap<K, V>>,
+    insert_notify: Arc<Notify>,
+    cache_metadata: Arc<RwLock<CacheMetadata<K>>>,
+    _cancellation_token_drop_guard: tokio_util::sync::DropGuard,
+    // User defined function to get the next key for a given key
+    // The function provides the key and the getter function to the DashMap in case a lookup is necessary
+    next_key_function:
+        Arc<dyn Fn(&K, &dyn Fn(&K) -> Option<V>) -> Option<K> + Send + Sync + 'static>,
```

Comment:
Why is this the right approach? Defining a function seems pretty complicated. A comment w/ example would be helpful.